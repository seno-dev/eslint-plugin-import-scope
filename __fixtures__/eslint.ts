import { readFile } from 'fs/promises'
import path from 'path'

import * as parser from '@typescript-eslint/parser'
import { TSESLint } from '@typescript-eslint/utils'
import { Program } from 'typescript'

import { rules } from './rules.js'
import importScope from '../src/index.js'

class ESLintTester {
  #projectRoot: string
  #linter: TSESLint.Linter
  #program: Program

  constructor(projectRoot: string) {
    this.#projectRoot = projectRoot
    this.#linter = new TSESLint.Linter({
      cwd: this.#projectRoot,
      configType: 'flat',
    })
    this.#program = parser.createProgram('./tsconfig.json', projectRoot)
  }

  async lintFile(filePath: string) {
    const fileAbsolutePath = path.join(this.#projectRoot, filePath)
    const code = await readFile(fileAbsolutePath, {
      encoding: 'utf8',
    })

    return this.#linter.verify(
      code,
      {
        files: ['**/*.ts'],
        languageOptions: {
          parser,
          parserOptions: {
            ecmaVersion: 2020,
            tsconfigRootDir: this.#projectRoot,
            project: './tsconfig.json',
            sourceType: 'module',
            program: this.#program,
          },
        },
        plugins: {
          'import-scope': importScope,
        },
        rules: {
          'import-scope/import-scope': ['error', rules],
        },
      },
      {
        filename: fileAbsolutePath,
      },
    )
  }
}

let cache: ESLintTester | undefined
/**
 * get an ESLint instance for testing.
 */
export function getESLintTester(): ESLintTester {
  const projectRoot = process.cwd()
  return (cache ||= new ESLintTester(projectRoot))
}
