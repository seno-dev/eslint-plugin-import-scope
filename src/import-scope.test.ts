import { join } from 'path'

import { getESLintTester } from '../__fixtures__/eslint.js'

const tester = getESLintTester()

describe('import-scope', () => {
  it.each([
    'app/page.ts',
    'app/a/page.ts',
    'features/a/api.ts',
    'features/b/api.ts',
    'features/a/utils.ts',
    'features/b/utils.ts',
  ])('%s', async (filename) => {
    const result = await tester.lintFile(join('__fixtures__', filename))
    expect(result).toMatchSnapshot()
  })
})
