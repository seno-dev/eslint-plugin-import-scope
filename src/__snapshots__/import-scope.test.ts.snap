// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import-scope > app/a/page.ts 1`] = `
[
  {
    "column": 1,
    "endColumn": 32,
    "endLine": 4,
    "line": 4,
    "message": "Importing '__fixtures__/features/b/utils.ts' is only allowed inside '__fixtures__/features/b'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 41,
    "endLine": 9,
    "line": 9,
    "message": "Importing '__fixtures__/features/b/utils.ts' is only allowed inside '__fixtures__/features/b'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
]
`;

exports[`import-scope > app/page.ts 1`] = `
[
  {
    "column": 1,
    "endColumn": 29,
    "endLine": 2,
    "line": 2,
    "message": "Importing '__fixtures__/features/a/utils.ts' is only allowed inside '__fixtures__/features/a' or '__fixtures__/app/a'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 29,
    "endLine": 4,
    "line": 4,
    "message": "Importing '__fixtures__/features/b/utils.ts' is only allowed inside '__fixtures__/features/b'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 41,
    "endLine": 7,
    "line": 7,
    "message": "Importing '__fixtures__/features/a/utils.ts' is only allowed inside '__fixtures__/features/a' or '__fixtures__/app/a'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 41,
    "endLine": 9,
    "line": 9,
    "message": "Importing '__fixtures__/features/b/utils.ts' is only allowed inside '__fixtures__/features/b'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
]
`;

exports[`import-scope > features/a/api.ts 1`] = `
[
  {
    "column": 1,
    "endColumn": 20,
    "endLine": 4,
    "line": 4,
    "message": "Importing '__fixtures__/features/b/utils.ts' is only allowed inside '__fixtures__/features/b'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 41,
    "endLine": 9,
    "line": 9,
    "message": "Importing '__fixtures__/features/b/utils.ts' is only allowed inside '__fixtures__/features/b'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
]
`;

exports[`import-scope > features/a/utils.ts 1`] = `[]`;

exports[`import-scope > features/b/api.ts 1`] = `
[
  {
    "column": 1,
    "endColumn": 20,
    "endLine": 2,
    "line": 2,
    "message": "Importing '__fixtures__/features/a/utils.ts' is only allowed inside '__fixtures__/features/a' or '__fixtures__/app/a'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 41,
    "endLine": 7,
    "line": 7,
    "message": "Importing '__fixtures__/features/a/utils.ts' is only allowed inside '__fixtures__/features/a' or '__fixtures__/app/a'",
    "messageId": "outside-scope",
    "nodeType": "ImportDeclaration",
    "ruleId": "import-scope/import-scope",
    "severity": 2,
  },
]
`;

exports[`import-scope > features/b/utils.ts 1`] = `[]`;
